app/views/course/assessment/answer/programming/_test_cases.html.slim

```diff
@@ -1,127 +1,45 @@
-              div.expected = format_html(test_case.expected)
-            - if can?(:grade, answer.answer)
-              td = simple_format(get_output(test_case_result))
-            - if @submission.attempting?
-              td
-                - if !test_case_result || !test_case_result.passed?
-                  = simple_format(get_hint(test_case, test_case_result))
-            td
-              - if test_case_result && test_case_result.passed?
-                = fa_icon 'check'.freeze
-              - elsif test_case_result
-                = fa_icon 'times'.freeze
+- if is_grader
+  - if test_cases_by_type.key?('private_test') || test_cases_by_type.key?('evaluation_test')
+    span.bg-danger.text-danger
+      = fa_icon 'exclamation-triangle'.freeze
+      =< t('.privacy_warning')
 
-- failed_public_test_cases = public_test_cases. \
-    select { |t| answer_test_results_hash[t] && !answer_test_results_hash[t].passed? }
-- if !failed_public_test_cases.empty?
-    - test_case = failed_public_test_cases.first
-    - test_case_result = answer_test_results_hash[test_case]
-    div.panel.panel-danger
-      div.panel-heading = t('.failed_public_message')
-      div.panel-body = simple_format(get_hint(test_case, test_case_result))
+  - if test_cases_and_results.key?('private_test')
+    = render partial: 'course/assessment/answer/programming/test_cases_of_type',
+      locals: { test_cases_and_results: test_cases_and_results['private_test'],
+                type_heading: t('.private'), is_grader: is_grader }
 
-- failed_private_test_cases = private_test_cases. \
-    select { |t| answer_test_results_hash[t] && !answer_test_results_hash[t].passed? }
-- if failed_public_test_cases.empty? && !failed_private_test_cases.empty?
-    - test_case = failed_private_test_cases.first
-    - test_case_result = answer_test_results_hash[test_case]
-    div.panel.panel-danger
-      div.panel-heading = t('.failed_private_message')
-      div.panel-body = simple_format(get_hint(test_case, test_case_result))
+  - if test_cases_and_results.key?('evaluation_test')
+    = render partial: 'course/assessment/answer/programming/test_cases_of_type',
+      locals: { test_cases_and_results: test_cases_and_results['evaluation_test'],
+                type_heading: t('.evaluation'), is_grader: is_grader }
 
-- failed_evaluation_test_cases = evaluation_test_cases. \
-    select { |t| answer_test_results_hash[t] && !answer_test_results_hash[t].passed? }
-- if can?(:grade, answer.answer) && !failed_evaluation_test_cases.empty?
-    - test_case = failed_evaluation_test_cases.first
-    - test_case_result = answer_test_results_hash[test_case]
-    div.panel.panel-danger
-      div.panel-heading = t('.failed_evaluation_message')
-      div.panel-body = simple_format(get_hint(test_case, test_case_result))
+/ Test failure warning panel display
+- failed_test_cases_by_type = get_failed_test_cases_by_type(test_cases_and_results)
+
+- if failed_test_cases_by_type['public_test']
```

Comment:
Probably not important to show evaluation test hints since only graders know about them and they can see the code anyway.

Hint might not even be set for that very reason.